Wikipedia/Code/ArticleFetcher.swift

```diff
@@ -0,0 +1,177 @@
+                return
+            }
+            
+            if let error = error {
+               completion(.failure(error))
+               return
+           }
+            
+            guard let urlStrings = urlStrings else {
+                completion(.failure(ArticleFetcherError.missingData))
+                return
+            }
+            
+            let result = urlStrings.map { (urlString) -> URL? in
+                let scheme = siteURL.scheme ?? "https"
+                let finalString = "\(scheme):\(urlString)"
+                return URL(string: finalString)
+            }.compactMap{ $0 }
+            
+            completion(.success(result))
+        }
+    }
+    
+    func fetchMediaListURLs(with url: URL, siteURL: URL, completion: @escaping (Result<[URL], Error>) -> Void) -> URLSessionTask? {
+        
+        return session.jsonDecodableTask(with: url) { (mediaList: MediaList?, response: URLResponse?, error: Error?) in
+            if let statusCode = (response as? HTTPURLResponse)?.statusCode,
+                statusCode == 404 {
+                completion(.failure(ArticleFetcherError.doesNotExist))
+                return
+            }
+            
+            if let error = error {
+               completion(.failure(error))
+               return
+           }
+            
+            guard let mediaList = mediaList else {
+                completion(.failure(ArticleFetcherError.missingData))
+                return
+            }
+            
+            let sources = mediaList.items.flatMap { (item) -> [MediaListItemSource] in
+                return item.sources
+            }
+            
+            let result = sources.map { (source) -> URL? in
+                let scheme = siteURL.scheme ?? "https"
+                let finalString = "\(scheme):\(source.urlString)"
+                return URL(string: finalString)
+            }.compactMap{ $0 }
+            
+            completion(.success(result))
+        }
+    }
+    
+    func handleDownloadTaskCompletion(url: URL, fileURL: URL?, response: URLResponse?, error: Error?, completion: @escaping DownloadCompletion) {
+        if let error = error {
+            completion(error, url, nil, nil)
+            return
+        }
+        guard let fileURL = fileURL, let response = response else {
+            completion(Fetcher.unexpectedResponseError, url, nil, nil)
+            return
+        }
+        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
+            completion(Fetcher.unexpectedResponseError, url, nil, nil)
+            return
+        }
+        completion(nil, url, fileURL, response.mimeType)
+    }
+}
+
+fileprivate struct MediaListItemSource: Codable {
+    let urlString: String
+    let scale: String
+    
+    enum CodingKeys: String, CodingKey {
+        case urlString = "src"
+        case scale
+    }
+}
+
+fileprivate enum MediaListItemType: String, Codable {
+    case image
```

Comment:
Is there a way to add a catch-all here?

Reply:
Tried an approach here: https://github.com/wikimedia/wikipedia-ios/pull/3408/commits/eb5def903cb40c7ef45690769c78a74ce1d9a02f

Reply:
@joewalsh will do, good catch